### ISSUE TITLE
Support JWE for OpenID Connect Client (RP)

### ISSUE THREAD

**Issue by teddyjtorres** [OpenLiberty/open-liberty#14646]
Support accepting JWE tokens in the OpenID Connect Relying Party implementation.

Section 10 and relevant subsections: https://openid.net/specs/openid-connect-core-1_0.html#SigEnc applied to the access_token and id_token.

This is a sister feature to [Support JWE for OpenID Connect Provider (OP) #16625](https://github.com/OpenLiberty/open-liberty/issues/16625).

When ready, add links to the Upcoming Feature Overview document and Feature Test Summary issue:
- UFO: https://ibm.box.com/s/t1wgxhsztpblrwt34e4vt16t3fc2ab5a
- FTS: https://github.com/OpenLiberty/open-liberty/issues/17035
- Beta Blog Post (if applicable): https://github.com/OpenLiberty/open-liberty/issues/18387
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/19629
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] No Design / No UFO Approval ([Arthur De Magalhaes](https://github.com/arthurdm) - cloud / [Alasdair Nottingham](https://github.com/NottyCode) - server) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point) - https://github.ibm.com/websphere/liberty-docs/issues/2410 and https://github.com/OpenLiberty/docs/issues/4730
- [x] Create a child task of this epic entitled "FAT Approval Test Summary". Add the link in above.

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [x] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([Adam Yoho](https://github.com/ayoho), [Martin Holder](https://github.com/mhldr), or [Dave Waddling](https://github.com/dave-waddling)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [x] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Kareen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [x] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [x] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [x] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [x] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [x] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by teddyjtorres** [comment]
Feature socialized. Pending Design Approval @NottyCode.

**Comment by NottyCode** [comment]
@teddyjtorres can we also update the oidcLogin feature of social login please.

**Comment by teddyjtorres** [comment]
@NottyCode Yes, we can update the oidcLogin element. Thank you for your review.

**Comment by teddyjtorres** [comment]
@NottyCode The UFO was updated.

**Comment by ayoho** [comment]
STE slides available: https://ibm.box.com/s/vjdizl1u1d8affx61vwxejsljspemlnd

**Comment by ayoho** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?
    > Yes, the Serviceability slide covers the error scenarios most likely to be encountered and notes that NLS error messages will be written to the server log. The implementation does, indeed, have NLS messages - both existing and new - that cover problems that might be encountered.

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?
    > We demoed the functionality to the Base Security squad, who were not involved in the development or testing of the feature. We demonstrated the normal golden path flow as well as a couple error scenarios (e.g. JWS sent when a JWE is expected, vice versa, the payload of the JWE is not a JWS token). We also discussed other error scenarios like signature and encryption algorithm mismatches that we have tests for, but which I didn't have tokens readily available to show. The group was convinced we had a good handle on the error scenarios and that we give customers enough information to fix problems as much as they could on their own without contacting support. Our FTS issue (#17035) contains extensive information about the scenarios we test in automation.

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
    > SVT was not necessary for this feature since all of the functionality could reasonably be tested in FATs.

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.
    > WAS L2: SEC
    > WAS L3: Security SSO

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?
    > No, no new metrics or JSON events emitted.

**Comment by samwatibm** [comment]
Move to 22.0.0.1 per discussion at scrum of scrums

**Comment by teddyjtorres** [comment]
@chirp1 @jhanders34 @hanczaryk Hi. Please let us know if anything else is needed to get approvals. We are targeting 22.0.0.1. Thanks.

**Comment by chirp1** [comment]
Teddy opened a doc issue at https://github.com/OpenLiberty/docs/issues/4730 and provided the information to document. Approving.

**Comment by teddyjtorres** [comment]
@jhanders34 Please let us know if anything else is needed to get approvals for performance. We are targeting 22.0.0.1. Thanks.
